### PR TITLE
feat: add native addon to `cphase` API

### PR DESCRIPTION
**Pull request by steff456**
Part of https://github.com/stdlib-js/todo/issues/1454

## Description

> What is the purpose of this pull request?

This pull request:

-   Adds support for Complex128 dtype
-   Adds C implementation
-   Adds native addon with benchmark, tests, examples and readme update

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   Part of https://github.com/stdlib-js/todo/issues/1454

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/math/base/special/cphase/README.md [hunk 1]
```diff
@@ -74,14 +76,119 @@ for ( i = 0; i < 100; i++ ) {
...
+<!-- /.notes -->
+
+<!-- C API usage examples. -->
+
+<section class="examples">
+
+### Examples
+
+```c
+#include "stdlib/math/base/special/cphase.h"
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/cphase/README.md:147]: This example is not correct, as `cphase` returns a scalar value.

File: lib/node_modules/@stdlib/math/base/special/cphase/README.md [hunk 2]
```diff
@@ -38,12 +38,14 @@ The [argument][complex-number-argument] of a complex number, also known as the *
 var cphase = require( '@stdlib/math/base/special/cphase' );
 ```
 
-#### cphase( re, im )
+#### cphase( z )
 
-Computes the [argument][complex-number-argument] of a `complex` number comprised of a **real** component `re` and an **imaginary** component `im`.
+Computes the [argument][complex-number-argument] of a double-precision complex floating-point number.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/cphase/README.md:43]: Can you also update the main README description to use the `double-precision` language.

File: lib/node_modules/@stdlib/math/base/special/cphase/docs/repl.txt [hunk 3]
```diff
@@ -1,5 +1,5 @@
 
-{{alias}}( re, im )
+{{alias}}( z )
     Computes the argument of a complex number in radians.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/cphase/docs/repl.txt]: Can you update the description here? Thanks!

File: lib/node_modules/@stdlib/math/base/special/cphase/docs/types/index.d.ts [hunk 4]
```diff
@@ -18,22 +18,27 @@
 
 // TypeScript Version: 2.0
 
+/// <reference types="@stdlib/types"/>
+
+import { Complex128 } from '@stdlib/types/object';
+
 /**
 * Computes the argument of a complex number in radians.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/cphase/docs/types/index.d.ts]: Can you update the description here? Thanks!

File: lib/node_modules/@stdlib/math/base/special/cphase/README.md (hunk 1 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/cphase/README.md:147]: See `example.c`.

File: lib/node_modules/@stdlib/math/base/special/cphase/README.md [hunk 5]
```diff
@@ -74,14 +76,119 @@ for ( i = 0; i < 100; i++ ) {
     re = round( randu()*100.0 ) - 50.0;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/cphase/README.md]: Can you replace with `random/base/uniform`?

File: lib/node_modules/@stdlib/math/base/special/cphase/examples/index.js [hunk 6]
```diff
@@ -21,8 +21,6 @@
 var Complex128 = require( '@stdlib/complex/float64' );
 var randu = require( '@stdlib/random/base/randu' );
 var round = require( '@stdlib/math/base/special/round' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/cphase/examples/index.js]: See comment in README.

File: lib/node_modules/@stdlib/math/base/special/cphase/include/stdlib/math/base/special/cphase.h [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+/*
+* If C++, prevent name mangling so that the compiler emits a binary file having undecorated names, thus mirroring the behavior of a C compiler.
+*/
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+* Computes the argument of a complex number in radians.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/cphase/include/stdlib/math/base/special/cphase.h]: Description.

File: lib/node_modules/@stdlib/math/base/special/cphase/lib/index.js [hunk 8]
```diff
@@ -24,9 +24,10 @@
 * @module @stdlib/math/base/special/cphase
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/cphase/lib/index.js:24]: Can you update the description here? Thanks!

File: lib/node_modules/@stdlib/math/base/special/cphase/lib/main.js [hunk 9]
```diff
@@ -21,23 +21,25 @@
...
 
 var atan2 = require( '@stdlib/math/base/special/atan2' );
+var real = require( '@stdlib/complex/real' );
+var imag = require( '@stdlib/complex/imag' );
 
 
 // MAIN //
 
 /**
 * Computes the argument of a complex number in radians.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/cphase/lib/main.js]: Desc.

File: lib/node_modules/@stdlib/math/base/special/cphase/lib/main.js [hunk 10]
```diff
@@ -21,23 +21,25 @@
...
 *
-* @param {number} re - real component
-* @param {number} im - imaginary component
+* @param {Complex128} z - complex number
 * @returns {number} argument
 *
 * @example
-* var phi = cphase( 5.0, 3.0 );
+* var Complex128 = require( '@stdlib/complex/float64' );
+* var phi = cphase( new Complex128( 5.0, 3.0 ) );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/cphase/lib/main.js:39]:
```suggestion
*
* var phi = cphase( new Complex128( 5.0, 3.0 ) );
```

File: lib/node_modules/@stdlib/math/base/special/cphase/lib/native.js [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+// MODULES //
+
+var addon = require( './../src/addon.node' );
+
+
+// MAIN //
+
+/**
+* Computes the argument of a complex number in radians.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/cphase/lib/native.js]: Desc.

File: lib/node_modules/@stdlib/math/base/special/cphase/lib/native.js [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+*
+* @private
+* @param {Complex128} z - complex number
+* @returns {number} argument
+*
+* @example
+* var Complex128 = require( '@stdlib/complex/float64' );
+* var phi = cphase( new Complex128( 5.0, 3.0 ) );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/cphase/lib/native.js:37]:
```suggestion
*
* var phi = cphase( new Complex128( 5.0, 3.0 ) );
```

Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/cphase/lib/native.js:37]: cc @Planeshifter We should be able to create a lint rule for this, correct?

@steff456 Would you mind opening an internal todo for this? Thanks!

File: lib/node_modules/@stdlib/math/base/special/cphase/src/main.c [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+* limitations under the License.
+*/
+
+#include "stdlib/math/base/special/cphase.h"
+#include "stdlib/complex/float64.h"
+#include "stdlib/complex/reim.h"
+#include <math.h>
+
+/**
+* Computes the argument of a complex number in radians.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/cphase/src/main.c]: Desc.

File: lib/node_modules/@stdlib/math/base/special/cphase/src/main.c [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+* stdlib_complex128_t z = stdlib_complex128( 5.0, 3.0 );
+* double out = stdlib_base_cphase( z );
+* // returns ~0.5404
+*/
+double stdlib_base_cphase( const stdlib_complex128_t z ) {
+	double re;
+	double im;
+
+	stdlib_reim( z, &re, &im );
+	return atan2( im, re );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/cphase/src/main.c]:
```suggestion
	return atan2( im, re ); // TODO: replace with stdlib function once available
```

File: lib/node_modules/@stdlib/math/base/special/cpolar/lib/cpolar.js [hunk 15]
```diff
@@ -47,7 +47,7 @@ var Complex128 = require( '@stdlib/complex/float64' );
 */
 function cpolar( out, re, im ) {
 	out[ 0 ] = cabs( new Complex128( re, im ) );
-	out[ 1 ] = cphase( re, im );
+	out[ 1 ] = cphase( new Complex128( re, im ) );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/cpolar/lib/cpolar.js]: You can avoid duplicate creation of the complex number.

**Review by kgryte — CHANGES_REQUESTED** [review]
Thanks! Left a few comments.

File: lib/node_modules/@stdlib/math/base/special/cphase/lib/native.js (hunk 12 above)
Comment (steff456) [reply lib/node_modules/@stdlib/math/base/special/cphase/lib/native.js:37]: For reference: https://github.com/stdlib-js/todo/issues/1657

File: lib/node_modules/@stdlib/math/base/special/cphase/README.md [hunk 16]
```diff
@@ -20,7 +20,7 @@ limitations under the License.
 
 # phase
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/cphase/README.md]:
```suggestion
# cphase
```

File: lib/node_modules/@stdlib/math/base/special/cphase/README.md [hunk 17]
```diff
@@ -59,29 +61,127 @@ var phi = cphase( 5.0, 3.0 );
...
+    double im1;
+    double re2;
+    double im2;
+    int i;
+    for ( i = 0; i < 4; i++ ) {
+        v = x[ i ];
+        y = stdlib_base_cphase( v );
+        stdlib_reim( v, &re1, &im1 );
+        stdlib_reim( y, &re2, &im2 );
+        printf( "fcphase(%lf + %lfi) = %lf\n", re, im, y );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/cphase/README.md]:
```suggestion
        printf( "cphase(%lf + %lfi) = %lf\n", re, im, y );
```

File: lib/node_modules/@stdlib/math/base/special/cphase/docs/types/test.ts [hunk 18]
```diff
@@ -16,38 +16,27 @@
...
 */
 
+import Complex128 = require( '@stdlib/complex/float64' );
 import cphase = require( './index' );
 
 
 // TESTS //
 
-// The function returns a number...
+// The function returns a double-precision complex floating-point number...
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/cphase/docs/types/test.ts]:
```suggestion
// The function returns a number...
```

**Review by kgryte — APPROVED** [review]
LGTM. Thanks, @steff456!

**Pull request merged**